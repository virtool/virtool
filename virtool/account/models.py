from datetime import datetime

from email_validator import EmailSyntaxError, validate_email
from pydantic import ConstrainedStr, validator

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.enums import AnalysisWorkflow
from virtool_core.models.group import GroupMinimal, Permissions
from virtool_core.models.user import User


def check_email(email: str | None) -> str | None:
    """Checks if the given email is valid."""
    if email is None:
        return None

    try:
        validate_email(email)
    except EmailSyntaxError:
        raise ValueError("The format of the email is invalid")

    return email


class AccountSettings(BaseModel):
    quick_analyze_workflow: AnalysisWorkflow
    show_ids: bool
    show_versions: bool
    skip_quick_analyze_dialog: bool


class ConstrainedEmail(ConstrainedStr):
    strip_whitespace = True


class Account(User):
    settings: AccountSettings
    email: ConstrainedEmail | None

    _email_validation = validator("email", allow_reuse=True)(check_email)


class APIKey(BaseModel):
    id: str
    created_at: datetime
    groups: list[GroupMinimal]
    name: str
    permissions: Permissions
