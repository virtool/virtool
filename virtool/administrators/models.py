from typing import List

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.roles import AdministratorRole
from virtool_core.models.user import UserNested


class AdministratorMinimal(UserNested):
    role: AdministratorRole


class Administrator(AdministratorMinimal):
    available_roles: List[dict]


class AdministratorSearch(BaseModel):
    items: List[AdministratorMinimal]
    available_roles: List[dict]
