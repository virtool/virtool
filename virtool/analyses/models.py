import enum
from sqlalchemy import Column, Integer, String, DateTime, Enum

from virtool.postgres import Base


class AnalysisFormat(str, enum.Enum):
    """
    Enumerated type for analysis file formats

    """

    sam = "sam"
    bam = "bam"
    fasta = "fasta"
    fastq = "fastq"
    csv = "csv"
    tsv = "tsv"
    json = "json"


class AnalysisFile(Base):
    """
    SQL model to store new analysis files

    """
    __tablename__ = "analysis_files"

    id = Column(Integer, primary_key=True)
    analysis = Column(String)
    description = Column(String)
    format = Column(Enum(AnalysisFormat))
    name = Column(String)
    name_on_disk = Column(String, unique=True)
    size = Column(Integer)
    uploaded_at = Column(DateTime)

    def __repr__(self):
        return f"<AnalysisFile(id={self.id}, analysis={self.analysis}, description={self.description}, " \
               f"format={self.format}, name={self.name}, name_on_disk={self.name_on_disk}, size={self.size}, " \
               f"uploaded_at={self.uploaded_at})>"
