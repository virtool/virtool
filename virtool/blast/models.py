from datetime import datetime

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.task import TaskNested


class NuvsBlast(BaseModel):
    id: int
    created_at: datetime
    updated_at: datetime
    last_checked_at: datetime
    error: str | None
    rid: str | None
    ready: bool = False
    result: dict | None
    task: TaskNested | None
