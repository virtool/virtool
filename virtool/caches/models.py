from sqlalchemy import Column, DateTime, Enum, Integer, String
from sqlalchemy.sql.schema import UniqueConstraint

from virtool.pg.utils import Base
from virtool.samples.models import ArtifactType


class SampleArtifactCache(Base):
    """
    SQL model to store a cached sample artifact

    """
    __tablename__ = "sample_artifacts_cache"
    __table_args__ = (UniqueConstraint("key", "name", "sample"),)

    id = Column(Integer, primary_key=True)
    key = Column(String, nullable=False)
    name = Column(String, nullable=False)
    name_on_disk = Column(String)
    sample = Column(String, nullable=False)
    size = Column(Integer)
    type = Column(Enum(ArtifactType), nullable=False)
    uploaded_at = Column(DateTime)


class SampleReadsCache(Base):
    """
    SQL model to store cached sample reads files

    """
    __tablename__ = "sample_reads_cache"
    __tableargs__ = (UniqueConstraint("key", "name", "sample"),)

    id = Column(Integer, primary_key=True)
    key = Column(String, nullable=False)
    name = Column(String(length=13), nullable=False)
    name_on_disk = Column(String, nullable=False)
    sample = Column(String, nullable=False)
    size = Column(Integer)
    uploaded_at = Column(DateTime)