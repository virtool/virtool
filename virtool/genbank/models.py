from pydantic import validator

from virtool_core.models.basemodel import BaseModel

VALID_SEQUENCE = frozenset("cagntCAGNT")


class Genbank(BaseModel):
    accession: str
    definition: str
    host: str
    sequence: str

    @validator("sequence")
    def check_sequence(cls, sequence: str | None) -> str | None:
        """
        Checks if the given sequence is valid.
        """
        if sequence:
            if not set(sequence) <= VALID_SEQUENCE:
                raise ValueError("The format of the sequence is invalid")

            return sequence.upper()
        return None
