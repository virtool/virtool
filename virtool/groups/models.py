from __future__ import annotations

from typing import TYPE_CHECKING, List

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.searchresult import SearchResult

if TYPE_CHECKING:
    from virtool_core.models.user import UserNested


class Permissions(BaseModel):
    """
    The permissions possessed by a user and group.
    """

    cancel_job: bool = False
    create_ref: bool = False
    create_sample: bool = False
    modify_hmm: bool = False
    modify_subtraction: bool = False
    remove_file: bool = False
    remove_job: bool = False
    upload_file: bool = False


class GroupMinimal(BaseModel):
    id: int | str
    legacy_id: str | None
    name: str


class Group(GroupMinimal):
    permissions: Permissions
    users: list[UserNested]


class GroupSearchResult(SearchResult):
    items: List[GroupMinimal]
