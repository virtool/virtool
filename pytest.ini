[pytest]
addopts = --aiohttp-loop uvloop --capture tee-sys
asyncio_mode = auto
