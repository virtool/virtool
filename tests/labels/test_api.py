import pytest
from virtool.labels.models import Label


async def test_find(spawn_client, test_session):
    """
    Test that a ``GET /api/labels`` return a complete list of labels.

    """
    client = await spawn_client(authorize=True, administrator=True)

    label_1 = Label(id=1, name="Bug", color="#a83432", description="This is a bug")
    label_2 = Label(id=2, name="Question", color="#03fc20", description="This is a question")

    await client.db.samples.insert_many([
        {
            "_id": "foo",
            "name": "Foo",
            "labels": [2]
        },
        {
            "_id": "bar",
            "name": "Bar",
            "labels": [1]
        },
        {
            "_id": "baz",
            "name": "Baz",
            "labels": [2]
        }
    ])

    async with test_session as session:
        session.add(label_1)
        session.add(label_2)
        await session.commit()

    resp = await client.get("/api/labels")
    assert resp.status == 200

    assert await resp.json() == [
        {
            "id": 1,
            "name": "Bug",
            "color": "#a83432",
            "description": "This is a bug",
            "count": 1
        },
        {
            "id": 2,
            "name": "Question",
            "color": "#03fc20",
            "description": "This is a question",
            "count": 2
        }
    ]


@pytest.mark.parametrize("error", [None, "404"])
async def test_get(error, spawn_client, all_permissions, test_session, resp_is):
    """
    Test that a ``GET /api/labels/:label_id`` return the correct label document.

    """
    client = await spawn_client(authorize=True, administrator=True)

    await client.db.samples.insert_many([
        {
            "_id": "foo",
            "name": "Foo",
            "labels": [2]
        },
        {
            "_id": "bar",
            "name": "Bar",
            "labels": [1]
        },
        {
            "_id": "baz",
            "name": "Baz",
            "labels": [2]
        }
    ])

    if not error:
        label = Label(id=1, name="Bug", color="#a83432", description="This is a test")
        async with test_session as session:
            session.add(label)
            await session.commit()

    resp = await client.get("/api/labels/1")

    if error:
        assert await resp_is.not_found(resp)
        return

    assert resp.status == 200

    assert await resp.json() == {
        "id": 1,
        "name": "Bug",
        "color": "#a83432",
        "description": "This is a test",
        "count": 1
    }


@pytest.mark.parametrize("error", [None, "400_exists", "422_color"])
async def test_create(error, spawn_client, test_random_alphanumeric, test_session, resp_is):
    """
    Test that a label can be added to the database at ``POST /api/labels``.

    """
    client = await spawn_client(authorize=True, administrator=True)

    await client.db.samples.insert_many([
        {
            "_id": "foo",
            "name": "Foo",
            "labels": [2]
        },
        {
            "_id": "bar",
            "name": "Bar",
            "labels": [1]
        },
        {
            "_id": "baz",
            "name": "Baz",
            "labels": [2]
        }
    ])

    if error == "400_exists":
        label = Label(id=1, name="Bug")
        async with test_session as session:
            session.add(label)
            await session.commit()

    data = {
        "name": "Bug",
        "color": "#a83432",
        "description": "This is a bug"
    }

    if error == "422_color":
        data["color"] = "#1234567"

    resp = await client.post("/api/labels", data)

    if error == "400_exists":
        assert await resp_is.bad_request(resp, "Label name already exists")
        return

    if error == "422_color":
        assert resp.status == 422
        return

    assert resp.status == 201

    assert await resp.json() == {
        "id": 1,
        "name": "Bug",
        "color": "#a83432",
        "description": "This is a bug",
        "count": 1
    }


@pytest.mark.parametrize("error", [None, "404", "400_exists", "422_color", "422_data"])
async def test_edit(error, spawn_client, test_session, resp_is):
    """
        Test that a label can be edited to the database at ``PATCH /api/labels/:label_id``.

    """
    client = await spawn_client(authorize=True, administrator=True)

    await client.db.samples.insert_many([
        {
            "_id": "foo",
            "name": "Foo",
            "labels": [2]
        },
        {
            "_id": "bar",
            "name": "Bar",
            "labels": [1]
        },
        {
            "_id": "baz",
            "name": "Baz",
            "labels": [2]
        }
    ])

    if error != "404":
        label_1 = Label(id=1, name="Bug", color="#a83432", description="This is a bug")
        label_2 = Label(id=2, name="Question", color="#03fc20", description="Question from a user")
        async with test_session as session:
            session.add(label_1)
            session.add(label_2)
            await session.commit()

    data = dict()

    if error != "422_data":
        data = {
            "name": "Bug",
            "color": "#fc5203",
            "description": "Need to be fixed"
        }

    if error == "400_exists":
        data["name"] = "Question"

    if error == "422_color":
        data["color"] = "#123bzp"

    resp = await client.patch("/api/labels/1", data=data)

    if error == "404":
        assert await resp_is.not_found(resp)
        return

    if error == "400_exists":
        assert await resp_is.bad_request(resp, "Label name already exists")
        return

    if error == "422_color" or error == "422_data":
        assert resp.status == 422
        return

    assert resp.status == 200
    assert await resp.json() == {
        "id": 1,
        "name": "Bug",
        "color": "#fc5203",
        "description": "Need to be fixed",
        "count": 1
    }


@pytest.mark.parametrize("error", [None, "400"])
async def test_remove(error, spawn_client, test_session, resp_is):
    """
        Test that a label can be deleted to the database at ``DELETE /api/labels/:label_id``.

    """
    client = await spawn_client(authorize=True, administrator=True)

    if not error:
        label = Label(id=1, name="Bug", color="#a83432", description="This is a bug")
        async with test_session as session:
            session.add(label)
            await session.commit()

    resp = await client.delete("/api/labels/1")

    if error:
        assert await resp_is.not_found(resp)
        return

    assert await resp_is.no_content(resp)
