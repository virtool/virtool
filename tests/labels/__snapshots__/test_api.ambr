# serializer version: 1
# name: TestFind.test_find
  list([
    dict({
      'color': '#955887',
      'count': 2,
      'description': 'Measure example sense peace economy.',
      'id': 2,
      'name': 'Help',
    }),
    dict({
      'color': '#629F70',
      'count': 2,
      'description': 'American whole magazine truth stop whose.',
      'id': 1,
      'name': 'Three',
    }),
  ])
# ---
# name: TestFind.test_find_by_name
  list([
    dict({
      'color': '#629F70',
      'count': 0,
      'description': 'American whole magazine truth stop whose.',
      'id': 1,
      'name': 'Three',
    }),
  ])
# ---
# name: TestFind.test_find_by_name.1
  list([
  ])
# ---
# name: test_edit[None]
  dict({
    'color': '#629F70',
    'count': 2,
    'description': 'American whole magazine truth stop whose.',
    'id': 1,
    'name': 'Summer',
  })
# ---
# name: test_get[200]
  dict({
    'color': '#629F70',
    'count': 2,
    'description': 'American whole magazine truth stop whose.',
    'id': 1,
    'name': 'Three',
  })
# ---
# name: test_get[404]
  dict({
    'id': 'not_found',
    'message': 'Not found',
  })
# ---
# name: test_remove[204]
  None
# ---
# name: test_remove[404]
  dict({
    'id': 'not_found',
    'message': 'Not found',
  })
# ---
